pkg/reconciler/apis/apibinding/apibinding_reconcile_permissionclaim.go

```diff
@@ -0,0 +1,302 @@
+// Permission claims are considered invalid when the identity hashes are mismatched, and when there is no dynamic informer
+// for the group resource.
+func (c *controller) reconcilePermissionClaims(ctx context.Context, apiBinding *apisv1alpha1.APIBinding) error {
+	identityMismatch := false
+	lc := logicalcluster.From(apiBinding)
+
+	// Get all the bound resources in the locicalcluster
+	bindings, err := c.listAPIBindings(lc)
+	if err != nil {
+		return err
+	}
+	grsToBoundResource := map[apisv1alpha1.GroupResource]apisv1alpha1.BoundAPIResource{}
+	for _, binding := range bindings {
+		for _, resource := range binding.Status.BoundResources {
+			grsToBoundResource[apisv1alpha1.GroupResource{Group: resource.Group, Resource: resource.Resource}] = resource
+		}
+	}
+
+	invalidClaims := []apisv1alpha1.PermissionClaim{}
+	// get the added permission claims if added, add are just requeue the object.
+	addedPermissionClaims := mapGVRToClaim{}
+	for _, acceptedPC := range apiBinding.Spec.AcceptedPermissionClaims {
+		found := false
+		for _, observedPC := range apiBinding.Status.ObservedAcceptedPermissionClaims {
+			if acceptedPC.Equal(observedPC) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			if acceptedPC.IdentityHash != "" {
+				if grsToBoundResource[acceptedPC.GroupResource].Schema.IdentityHash != acceptedPC.IdentityHash {
+					identityMismatch = true
+					invalidClaims = append(invalidClaims, acceptedPC)
+					continue
+				}
+			}
+			claim, gvr, err := c.createPermissionClaimHelper(acceptedPC)
+			if err != nil {
+				return err
+			}
+			addedPermissionClaims[gvr] = append(addedPermissionClaims[gvr], claim)
+		}
+	}
+
+	removedPermissionClaims := mapGVRToClaim{}
+	for _, observedPC := range apiBinding.Status.ObservedAcceptedPermissionClaims {
+		found := false
+		for _, acceptedPC := range apiBinding.Spec.AcceptedPermissionClaims {
+			if observedPC.Equal(acceptedPC) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			claim, gvr, err := c.createPermissionClaimHelper(observedPC)
+			if err != nil {
+				return err
+			}
+			claims := removedPermissionClaims[gvr]
+			if claims == nil {
+				claims = []permissionClaimHelper{}
+			}
+			claims = append(claims, claim)
+			removedPermissionClaims[gvr] = claims
+		}
+	}
+
+	klog.V(2).Infof("permission controller adding permission claims: %v, removing: %v and found invalid claims: %s", addedPermissionClaims, removedPermissionClaims, invalidClaims)
+
+	for gvr, claims := range addedPermissionClaims {
+		for _, claim := range claims {
+			//TODO: use indexer instead of lister.
+			requirement, err := labels.NewRequirement(claim.key, selection.DoesNotExist, []string{claim.label})
+			if err != nil {
+				return err
+			}
+			objs, err := claim.lister.Lister().List(labels.Everything().Add(*requirement))
+			if err != nil {
+				return err
+			}
+			for _, obj := range objs {
+				err = c.patchGenericObject(ctx, obj, claim, gvr, lc)
+				if err != nil {
+					return err
```

Comment:
this means that one e.g. admission webhook of one resource will block the whole mechanism. Not good.

Reply:
Block in the sense that admission failing this update would wedge the controller? Since the VirtualWorkspace just keys off of the labels, you should still be able to go through it to see the ones that have been labelled.

Reply:
updated